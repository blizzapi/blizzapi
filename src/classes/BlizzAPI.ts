import * as helpers from '../helpers';
import * as utils from '../utils';
import { BattleNetAPI } from './abstract/BattleNetAPI';
import {
  Endpoint,
  BlizzAPIOptions,
  AccessTokenOptions,
  QueryOptions,
} from '../types';

export class BlizzAPI extends BattleNetAPI {
  readonly options: AccessTokenOptions;

  constructor(options: BlizzAPIOptions) {
    super({
      region: options.region,
      clientId: options.clientId,
      clientSecret: options.clientSecret,
      accessToken: options.accessToken,
    });
    this.options = {
      validateAccessTokenOnEachQuery: options.validateAccessTokenOnEachQuery || false,
      refreshExpiredAccessToken: options.refreshExpiredAccessToken || false,
      onAccessTokenExpired: options.onAccessTokenExpired || undefined,
      onAccessTokenRefresh: options.onAccessTokenRefresh || undefined,
    };
  }

  query = async (endpoint: Endpoint, options?: QueryOptions) =>
    helpers.query({
      endpoint,
      region: this.region,
      clientId: this.clientId,
      clientSecret: this.clientSecret,
      accessToken: await this.getAccessToken(),
      options: {
        ...this.options,
        ...options,
      },
    });

  static getAllRegions = utils.getAllRegions;

  static getAllRegionIds = utils.getAllRegionIds;

  static getAllRegionNames = utils.getAllRegionNames;

  static getRegionNameById = utils.getRegionNameById;

  static validateRegionId = utils.validateRegionId;

  static getRegionIdByName = utils.getRegionIdByName;

  static validateRegionName = utils.validateRegionName;

  static getAllLocales = utils.getAllLocales;

  static getAllLocaleNames = utils.getAllLocaleNames;

  static getLocalesByRegionId = utils.getLocalesByRegionId;

  static checkIfLocaleLooksValid = utils.checkIfLocaleLooksValid;

  static validateLocale = utils.validateLocale;

  static isLocaleValidForRegionId = utils.isLocaleValidForRegionId;

  static getAllSc2Realms = utils.getAllSc2Realms;

  static getAllAvailableSc2Realms = utils.getAllAvailableSc2Realms;

  static getSc2RealmsByRegionId = utils.getSc2RealmsByRegionId;

  static checkIfSc2RealmLooksValid = utils.checkIfSc2RealmLooksValid;

  static validateSc2Realm = utils.validateSc2Realm;

  static isSc2RealmValidForRegionId = utils.isSc2RealmValidForRegionId;

  static getDefaultLocaleNameForRegionId = utils.getDefaultLocaleNameForRegionId;

  static getAllDefaultLocaleNames = utils.getAllDefaultLocaleNames;
}
