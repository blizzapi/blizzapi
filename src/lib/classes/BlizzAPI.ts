import helpers from '../helpers';
import { Endpoint, RegionIdOrName } from '../../@types';
export interface BlizzAPIOptions {
  region: RegionIdOrName;
  clientId?: string;
  clientSecret?: string;
  accessToken?: string;
  validateAccessTokenOnEachQuery?: boolean;
  refreshExpiredAccessToken?: boolean;
  onAccessTokenExpire?: Function | undefined;
  onAccessTokenRevalidate?: Function | undefined;
}
export default class BlizzAPI {
  readonly options: BlizzAPIOptions;

  constructor(options: BlizzAPIOptions) {
    this.options = {
      region: options.region,
      validateAccessTokenOnEachQuery: options.validateAccessTokenOnEachQuery || false, // revalidate access token on each single query
      refreshExpiredAccessToken: options.refreshExpiredAccessToken || false, // revalidate access token if error 403
      onAccessTokenExpire: options.onAccessTokenExpire || undefined,
      onAccessTokenRevalidate: options.onAccessTokenRevalidate || undefined,
    }
  }

  // getAccessToken

  query = (endpoint: Endpoint) => helpers.query(endpoint);

  // util functions
}