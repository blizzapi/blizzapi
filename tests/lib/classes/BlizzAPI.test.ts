import BlizzAPI from '../../../src/classes/BlizzAPI';

describe('BlizzAPI class', () => {
  it('should be defined', () => {
    expect.assertions(1);
    expect(BlizzAPI).toBeDefined();
  });

  it('should define query method', () => {
    expect.assertions(1);
    expect(new BlizzAPI({
      region: '',
      clientId: '',
      clientSecret: '',
    }).query).toBeDefined();
  });

  it('should define getAccessToken method', () => {
    expect.assertions(1);
    expect(new BlizzAPI({
      region: '',
      clientId: '',
      clientSecret: '',
    }).getAccessToken).toBeDefined();
  });

  it('should define validateAccessToken static method', () => {
    expect(BlizzAPI.validateAccessToken).toBeDefined();
  });

  it('should define getAllRegions static method', () => {
    expect(BlizzAPI.getAllRegions).toBeDefined();
  });

  it('should define getAllRegionIds static method', () => {
    expect(BlizzAPI.getAllRegionIds).toBeDefined();
  });

  it('should define getAllRegionNames static method', () => {
    expect.assertions(1);
    expect(BlizzAPI.getAllRegionNames).toBeDefined();
  });

  it('should define getRegionNameById static method', () => {
    expect.assertions(1);
    expect(BlizzAPI.getRegionNameById).toBeDefined();
  });

  it('should define validateRegionId static method', () => {
    expect(BlizzAPI.validateRegionId).toBeDefined();
  });

  it('should define getRegionIdByName static method', () => {
    expect(BlizzAPI.getRegionIdByName).toBeDefined();
  });

  it('should define validateRegionName static method', () => {
    expect(BlizzAPI.validateRegionName).toBeDefined();
  });

  it('should define getAllLocales static method', () => {
    expect(BlizzAPI.getAllLocales).toBeDefined();
  });

  it('should define getAllLocaleNames static method', () => {
    expect(BlizzAPI.getAllLocaleNames).toBeDefined();
  });

  it('should define getLocalesByRegionId static method', () => {
    expect(BlizzAPI.getLocalesByRegionId).toBeDefined();
  });

  it('should define checkIfLocaleLooksValid static method', () => {
    expect(BlizzAPI.checkIfLocaleLooksValid).toBeDefined();
  });

  it('should define validateLocale static method', () => {
    expect(BlizzAPI.validateLocale).toBeDefined();
  });

  it('should define isLocaleValidForRegionId static method', () => {
    expect(BlizzAPI.isLocaleValidForRegionId).toBeDefined();
  });

  it('should define getAllSc2Realms static method', () => {
    expect(BlizzAPI.isLocaleValidForRegionId).toBeDefined();
  });

  it('should define getAllAvailableSc2Realms static method', () => {
    expect(BlizzAPI.getAllAvailableSc2Realms).toBeDefined();
  });

  it('should define getSc2RealmsByRegionId static method', () => {
    expect(BlizzAPI.getSc2RealmsByRegionId).toBeDefined();
  });

  it('should define checkIfSc2RealmLooksValid static method', () => {
    expect(BlizzAPI.checkIfSc2RealmLooksValid).toBeDefined();
  });

  it('should define validateSc2Realm static method', () => {
    expect(BlizzAPI.validateSc2Realm).toBeDefined();
  });

  it('should define isSc2RealmValidForRegionId static method', () => {
    expect(BlizzAPI.isSc2RealmValidForRegionId).toBeDefined();
  });

  it('should define getDefaultLocaleNameForRegionId static method', () => {
    expect(BlizzAPI.getDefaultLocaleNameForRegionId).toBeDefined();
  });

  it('should define getAllDefaultLocaleNames static method', () => {
    expect(BlizzAPI.getAllDefaultLocaleNames).toBeDefined();
  });

  it('should query an endpoint', async () => {
    expect.assertions(1);
    const bnetApi = new BlizzAPI({
      region: 'us',
      clientId: 'valid_client_id',
      clientSecret: 'valid_client_secret',
    });

    const response = await bnetApi.query('/sample/endpoint');
    expect(response).toMatchSnapshot();
  });

  it('should validate access token and return true', async () => {
    expect.assertions(1);
    const response = await BlizzAPI.validateAccessToken('us', 'valid_access_token');
    expect(response).toMatchSnapshot();
  });
});
