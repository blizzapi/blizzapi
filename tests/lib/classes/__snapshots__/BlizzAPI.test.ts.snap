// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BlizzAPI class should query an endpoint 1`] = `
Object {
  "access_token": "no_access_token_provided",
  "headers": Object {
    "Authorization": "Bearer no_access_token_provided",
  },
  "params": Object {
    "data": "default sample params",
  },
  "timeout": 10000,
  "title": "sample mocked data for GET request to https://us.api.blizzard.com/sample/endpoint",
}
`;

exports[`BlizzAPI class should validate access token and return true 1`] = `true`;
